# Distribuição Binomial

## Em um sistema de comunicação digital, cada pacote transmitido tem probabilidade de erro igual a 0.08.

### Considere o envio de 20 pacotes.

a) Determine a probabilidade de ocorrerem exatamente 3 erros.  
b) Determine a probabilidade de ocorrerem no máximo 2 erros.  
c) Determine a probabilidade de ocorrerem mais de 5 erros.  
d) Calcule média e variância da distribuição.  

In [ ]:
!pip install numpy
!pip install math

In [1]:
"""
Distribuição Binomial

Este script resolve os incisos (a), (b), (c) e (d) para uma variável aleatória
Binomial com:

n = 20 tentativas
p = 0.08 (probabilidade de erro)

São calculados:
- Valores teóricos
- Simulação via método da inversa
"""

import numpy as np
import math

# ==============================
# Parâmetros
# ==============================
p = 0.08           # probabilidade de erro
q = 1 - p          # probabilidade de acerto
n = 20             # número de tentativas
N = 100000         # número de amostras

# ==============================
# Função PMF Binomial (teórica)
# ==============================
def binomial_pmf(k, n, p):
    return math.comb(n, k) * (p**k) * ((1-p)**(n-k))

# ==============================
# Simulação (método da inversa)
# ==============================
def gerar_binomial(n, p, N):
    q = 1 - p
    c = p / q
    amostras = []
    u = np.random.uniform(0, 1, N)

    for ix in u:
        i = 0
        pr = q**n
        F = pr

        while ix >= F:
            pr = (c * (n - i) / (i + 1)) * pr
            F += pr
            i += 1

        amostras.append(i)

    return np.array(amostras)

# Gerar amostras
amostras = gerar_binomial(n, p, N)

# ==============================
# a) P(X = 3)
# ==============================
p_teorico_a = binomial_pmf(3, n, p)
p_simulado_a = np.mean(amostras == 3)

print("a) P(X = 3)")
print("Teórico:", p_teorico_a)
print("Simulado:", p_simulado_a)

# ==============================
# b) P(X ≤ 2)
# ==============================
p_teorico_b = sum(binomial_pmf(k, n, p) for k in range(3))
p_simulado_b = np.mean(amostras <= 2)

print("\nb) P(X ≤ 2)")
print("Teórico:", p_teorico_b)
print("Simulado:", p_simulado_b)

# ==============================
# c) P(X > 5)
# ==============================
p_teorico_c = 1 - sum(binomial_pmf(k, n, p) for k in range(6))
p_simulado_c = np.mean(amostras > 5)

print("\nc) P(X > 5)")
print("Teórico:", p_teorico_c)
print("Simulado:", p_simulado_c)

# ==============================
# d) Média e variância
# ==============================
# Teórico
media_teorica = n * p
variancia_teorica = n * p * (1 - p)

# Simulado
media_simulada = np.mean(amostras)
variancia_simulada = np.var(amostras)

print("\nd) Média e variância")
print("Média teórica:", media_teorica)
print("Média simulada:", media_simulada)
print("Variância teórica:", variancia_teorica)
print("Variância simulada:", variancia_simulada)


a) P(X = 3)
Teórico: 0.14143857663883505
Simulado: 0.14173

b) P(X ≤ 2)
Teórico: 0.7879462459747031
Simulado: 0.78781

c) P(X > 5)
Teórico: 0.0037994868343921295
Simulado: 0.00397

d) Média e variância
Média teórica: 1.6
Média simulada: 1.60191
Variância teórica: 1.4720000000000002
Variância simulada: 1.4722343518999998


## Resultados  
---  

### a) Probabilidade de exatamente 3 erros

\[
P(X = 3)
\]

- Teórico: 0.14143857663883505  
- Simulado: 0.14173  

---

### b) Probabilidade de no máximo 2 erros

\[
P(X ≤ 2)
\]

- Teórico: 0.7879462459747031  
- Simulado: 0.78781  

---

### c) Probabilidade de mais de 5 erros

\[
P(X > 5)
\]

- Teórico: 0.0037994868343921295  
- Simulado: 0.00397  

---

### d) Média e variância

- Média teórica: 1.6  
- Média simulada: 1.60191  

- Variância teórica: 1.4720000000000002  
- Variância simulada: 1.4722343518999998  
